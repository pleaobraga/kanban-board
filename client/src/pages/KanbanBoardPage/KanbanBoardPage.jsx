import './KanbanBoardPage.scss'

import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { DragDropContext } from 'react-beautiful-dnd'
import { useParams } from 'react-router-dom'

import Board from '../../components/Board'
import BoardHeader from '../../components/BoardHeader'
import Loading from '../../components/Loader'
import ErrorPage from '../ErrorPage'
import {
  getBoard as actionGetBoard,
  putBoard
} from '../../reducers/BoardReducer'
import { findObject, removeObjFromArray } from '../../utils/utils'

const KanbanBoardPage = () => {
  const dispatch = useDispatch()
  const pageContent = useSelector(state => state.board)
  const { boardName } = useParams()

  React.useEffect(() => {
    dispatch(actionGetBoard(boardName || 0))
  }, [])

  const changeCards = (
    listInitial,
    initialIndex,
    listDestination,
    destinationIndex
  ) => {
    const newInitialCards = Array.from(listInitial.Cards)
    const removedItem = newInitialCards.splice(initialIndex, 1)
    let newDestinationList = null

    let newlists = Array.from(pageContent.board.TaskLists)
    removeObjFromArray(newlists, listInitial)

    if (listInitial === listDestination) {
      newInitialCards.splice(destinationIndex, 0, removedItem[0])
    } else {
      const newDestinationCards = Array.from(listDestination.Cards)
      newDestinationCards.splice(destinationIndex, 0, removedItem[0])

      newDestinationList = {
        ...listDestination,
        Cards: newDestinationCards
      }

      removeObjFromArray(newlists, listDestination)
      newlists.push(newDestinationList)
    }

    const newlistInitial = {
      ...listInitial,
      Cards: newInitialCards
    }

    const sortList = [...newlists, newlistInitial].sort(
      (a, b) => a.index - b.index
    )

    dispatch(
      putBoard({
        board: { TaskLists: [...sortList] },
        card: {
          id: removedItem[0].id,
          cardIndex: destinationIndex,
          taskListId: listDestination.id
        }
      })
    )
  }

  const onDragEnd = result => {
    const { destination, source } = result

    if (!destination) {
      return
    }

    if (
      destination.droppableId === source.droppableId &&
      destination.index === source.index
    ) {
      return
    }

    const listInitial = findObject(
      pageContent.board.TaskLists,
      'id',
      source.droppableId
    )

    const listDestination = findObject(
      pageContent.board.TaskLists,
      'id',
      destination.droppableId
    )

    changeCards(listInitial, source.index, listDestination, destination.index)
  }

  const { errorContent, board } = pageContent

  return errorContent ? (
    <ErrorPage />
  ) : board ? (
    <div className="page page-kanban-board">
      <BoardHeader name={board.name} />
      <DragDropContext onDragEnd={onDragEnd}>
        <Board TaskLists={board.TaskLists} />
      </DragDropContext>
    </div>
  ) : (
    <Loading />
  )
}

export default KanbanBoardPage
