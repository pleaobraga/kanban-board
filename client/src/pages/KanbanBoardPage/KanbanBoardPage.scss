.page {
  &-kanban-board {
    align-items: flex-start;
  }
}
