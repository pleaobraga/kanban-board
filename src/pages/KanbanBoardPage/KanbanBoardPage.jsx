import './KanbanBoardPage.scss'

import React from 'react'
import { DragDropContext } from 'react-beautiful-dnd'

import Board from '../../components/Board'
import { findObject, removeObjFromArray } from '../../utils/utils'

class KanbanBoardPage extends React.Component {
  state = {
    lists: [
      {
        id: '0',
        title: 'Backlog',
        cards: [
          {
            id: 'test1',
            type: 'feature',
            duration: 1,
            severity: 'hight'
          },
          {
            id: 'test2',
            type: 'research',
            duration: 2,
            severity: 'medium'
          },
          {
            id: 'test3',
            type: 'content',
            duration: 3,
            severity: 'low'
          }
        ]
      },
      {
        id: '1',
        title: 'To DO',
        cards: []
      }
    ]
  }

  onDragEnd = result => {
    const { destination, source } = result

    if (!destination) {
      return
    }

    if (
      destination.droppableId === source.droppableId &&
      destination.index === source.index
    ) {
      return
    }

    const startColumn = findObject(this.state.lists, 'id', source.droppableId)
    const finishColum = findObject(
      this.state.lists,
      'id',
      destination.droppableId
    )

    if (startColumn.id === finishColum.id) {
      const column = findObject(this.state.lists, 'id', source.droppableId)
      const newCardsList = Array.from(column.cards)
      const removedItem = newCardsList.splice(source.index, 1)
      newCardsList.splice(destination.index, 0, removedItem[0])

      const newColumn = {
        ...column,
        cards: newCardsList
      }

      const newlists = removeObjFromArray(Array.from(this.state.lists), column)

      const sortList = [...newlists, newColumn].sort((a, b) => a.id - b.id)

      this.setState({
        lists: [...sortList]
      })
    } else {
      const startCards = Array.from(startColumn.cards)

      const removedItem = startCards.splice(source.index, 1)

      const newStart = {
        ...startColumn,
        cards: startCards
      }

      const finishCards = Array.from(finishColum.cards)

      finishCards.splice(destination.index, 0, removedItem[0])

      const newFinish = {
        ...finishColum,
        cards: finishCards
      }

      let newlists = Array.from(this.state.lists)

      removeObjFromArray(newlists, startColumn)
      removeObjFromArray(newlists, finishColum)

      const sortList = [...newlists, newStart, newFinish].sort(
        (a, b) => a.id - b.id
      )

      this.setState({
        lists: [...sortList]
      })
    }
  }

  render() {
    return (
      <div className="page page-kanban-board">
        <DragDropContext onDragEnd={this.onDragEnd}>
          <Board lists={this.state.lists} />
        </DragDropContext>
      </div>
    )
  }
}

export default KanbanBoardPage
