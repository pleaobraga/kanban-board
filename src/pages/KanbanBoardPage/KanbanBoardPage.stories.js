import React from 'react'
import { storiesOf } from '@storybook/react'
import KanbanBoardPage from './KanbanBoardPage.stories'

export default storiesOf('Pages | Kanban Board Page', module).add(
  'default',
  () => <KanbanBoardPage />,
  { info: { inline: true, header: false } }
)
