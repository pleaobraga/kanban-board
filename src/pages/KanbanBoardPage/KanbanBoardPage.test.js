import React from 'react'
import { shallow } from 'enzyme'
import KanbanBoardPage from './KanbanBoardPage'

describe('Kanban Board Page', () => {
  const welcomePage = shallow(<KanbanBoardPage />)

  it('render proprely', () => {
    expect(welcomePage).toMatchSnapshot()
  })
})
