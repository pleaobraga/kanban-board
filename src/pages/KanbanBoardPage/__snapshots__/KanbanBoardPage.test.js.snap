// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kanban Board Page render proprely 1`] = `
<div
  className="page page-kanban-board"
>
  <Board
    lists={
      Array [
        Object {
          "cards": Array [
            Object {
              "duration": 1,
              "severity": "hight",
              "type": "feature",
            },
            Object {
              "duration": 2,
              "severity": "medium",
              "type": "research",
            },
            Object {
              "duration": 3,
              "severity": "low",
              "type": "update",
            },
          ],
          "title": "Backlog",
        },
        Object {
          "cards": Array [
            Object {
              "duration": 1,
              "severity": "hight",
              "type": "feature",
            },
            Object {
              "duration": 2,
              "severity": "medium",
              "type": "research",
            },
            Object {
              "duration": 3,
              "severity": "low",
              "type": "update",
            },
          ],
          "title": "To do ",
        },
        Object {
          "cards": Array [
            Object {
              "duration": 1,
              "severity": "hight",
              "type": "feature",
            },
            Object {
              "duration": 2,
              "severity": "medium",
              "type": "research",
            },
            Object {
              "duration": 3,
              "severity": "low",
              "type": "update",
            },
          ],
          "title": "Done",
        },
      ]
    }
  />
</div>
`;
